### ISSUE TITLE
Should use `css()` function for wrapping styles.

### ISSUE THREAD

**Issue by alanbsmith** [Decisiv/styled-components-modifiers#11]
This is an optimization suggested by the styled-components team.

**Comment by andrewtpoe** [comment]
For reference, [this is the `css` function](https://www.styled-components.com/docs/api#css) being discussed.

What are your thoughts on this as a solution?

1. If a modifier is constructed so it either is a string or returns a string we call `css(modifierStyleString)` to ensure the css optimizations.
```
  // modifier is a string
  key: `
    // ...styles
  `
  // modifier returns a string
  key: () => `
    // ...styles
  `
```
1. If a modifier returns a function, we call it & use the result directly:
```
  key: () => css`
    // ...styles
  `
```
My concern with this approach is the following condition:
```
  key: css`
    // ...styles
  `
```
This will need some experimentation to determine if the `css` function from `styled-components` returns a string or a different datatype. If it returns just a string (specifically when there are no interpolations), my proposed solution would essentially result in the `css` function being called twice.

The preference will be for developers to use `css` when writing the modifiers as this will enable other tools like stylelint and possibly prettier.

**Comment by eugeneross** [comment]
Any updates on this?

**Comment by andrewtpoe** [comment]
At present, no. If you'd like to use the `css` helper now you can return an object in the styled components modifiers:

```
  modifierName: () => ({ 
    styles: css`
      // your modifying styles
    `,
  })
```

This works fine in v1, although our plan is to deprecate the returned object pattern in v2.

**Comment by eugeneross** [comment]
@andrewtpoe Oh okay, thanks! I'll keep an eye on this.

On a side note, I have a couple questions for what you guys do at Decisiv. Would it be okay to ping you via email with the questions I have? 😄

**Comment by mcmillion** [comment]
I've been looking for a better way to test conditional modifiers on styled components without being coupled directly to the CSS in said modifiers (with something like `.toHaveStyleRule()`). I've gone down a few rabbit holes (trying to add modifiers to the components `displayName`, etc). One idea I thought of was to use reusable chunks of CSS (using `css`) and then writing some helpers to assert that those styles were being used when a given modifier was expected. The issue I ran into relates to this topic I think.

Currently, it appears that the way props are applied to the CSS used by `applyStyleModifiers` differs from how styled components does it. For example, the following won't work:

```
const ButtonPrimary = css`
  background: ${props => props.theme.primaryColor};
`;

const BUTTON_MODIFIERS = {
  primary: () => `${ButtonPrimary}`,
}
```

I've tried various other ways, but I haven't found any that work. It seems that the way props are handled by `applyStyleModifiers` is different than how they're actually handled inside styled components (which makes sense given all of the examples use `${props.whatever}` rather than `{props => props.whatever}`.

Seems like this would be a breaking change to correct the behavior, however.

**Comment by andrewtpoe** [comment]
We have two utilities that we've built internally that help use internally to help with both building and testing styled components + modifiers. 

Basically, we add the modifiers as a property to the styled component, then iterate over the component with a modifier applied. On each iteration, we take a snapshot of the component. We don't use snapshots for much other than this, but for verifying styles are consistent we've found it to be a good fit.

Our exact utilities are not open sourced, but you can see how the `buildStyledComponent` util [here](https://github.com/andrewtpoe/react-companies/blob/master/src/utils/buildStyledComponent/index.js) works. It's very similar.

**Comment by andrewtpoe** [comment]
This issue was resolved in #20.
